--- src/main/export_passage.cc.orig	2010-03-13 21:59:08.000000000 -0500
+++ src/main/export_passage.cc	2010-05-27 12:55:56.968451561 -0400
@@ -37,7 +37,7 @@
 #include "main/sword.h"
 
 
-#define HTML_START "<HTML><HEAD><META HTTP-EQUIV=\"content-type\" CONTENT=\"text/html; CHARSET=utf-8\"><STYLE TYPE=\"text/css\"><!-- A { text-decoration:none } %s --></STYLE></HEAD><BODY>"
+#define HTML_START "<html><head><meta http-equiv=\"content-type\" content=\"text/html; charset=utf-8\"><style type=\"text/css\"><!-- A { text-decoration:none } %s --></style></head><body>"
 
 enum {
         TARGET_HTML,
--- src/gnome2/utilities.c.orig	2010-03-19 13:01:57.000000000 -0400
+++ src/gnome2/utilities.c	2010-05-27 12:55:56.969460768 -0400
@@ -1202,6 +1202,33 @@
 		gtk_html_set_editable(html, FALSE);
 #endif
 
+#ifdef USE_GTKMOZEMBED
+	// EVIL EVIL EVIL EVIL.
+	// crazy nonsense with xulrunner 1.9.2.3, failure to jump to anchor.
+	// force the issue by stuffing a javascript snippet inside <head></head>.
+	// there are forms of evil so dark that they should not be contemplated.
+	if (anchor || settings.special_anchor) {
+		gchar *buf;
+
+		// first, scribble out everything up to the closing </head>.
+		buf = strstr(text, "</head>");	// yes, lowercase.
+		assert(buf != NULL);	// don't be so stupid as not to include <head></head>.
+		offset = buf - text;
+		gecko_html_write(html, text, offset);
+		len -= offset;
+
+		// now write the javascript snippet.
+		buf = g_strdup_printf(
+		    "<script type=\"text/javascript\" language=\"javascript\">"
+		    " window.onload = function () { window.location.hash = \"%s\"; }"
+		    " </script>", (settings.special_anchor
+				   ? settings.special_anchor
+				   : anchor));
+		gecko_html_write(html, buf, strlen(buf));
+		g_free(buf);
+	}
+#endif /* USE_GTKMOZEMBED */
+
 	// html widgets are uptight about being handed
 	// huge quantities of text -- producer/consumer problem,
 	// and we mustn't overload the receiver.  10k chunks.
